target: add EFI support to armvirt

mcbridematt opened the pull request:
**2023-02-17: More platforms supported. See https://github.com/openwrt/openwrt/pull/4996#issuecomment-1433894604**
**2023-04-27: Improved handling of ESP "boot" partition and shared media (e.g sdcard boot) setups https://github.com/openwrt/openwrt/pull/4996#issuecomment-1524683095**
**2023-06-01: Update for kernel 6.1: https://github.com/openwrt/openwrt/pull/4996#issuecomment-1571438903**

As suggested in https://github.com/openwrt/openwrt/pull/4956, this adds EFI support to the existing "armvirt" target.

To keep the target Makefiles simple, EFI support for 32-bit ARM has been implemented as well, despite such systems being quite rare. You can still use the generated kernels and rootfs under QEMU directly without a boot firmware.

This patchset changes armvirt so it generates images for ARM machines with EFI support, as specified by the [Arm SystemReady](https://github.com/ARM-software/arm-systemready) standards. These systems behave a lot like existing x86 machines, e.g using EFI+GRUB to boot. This target does use some of the same code and scripts as x86 (especially around the GRUB EFI setup).

Versions of this target have been submitted in the past but didn't get any traction. Now that such machines are now more abundant I think now is a good time. We have been "shipping" this as the OpenWrt port for our Ten64 product for a while, though some Ten64-specific bits are not in this set.

The initial version has been tested with:

Traverse Ten64 (NXP LS1088A)
QEMU virt with both EDKII and u-boot.
Amazon EC2 Graviton instances (both virtual and bare-metal instances)
VMware Fusion for Apple Silicon (tech preview version).

The following changes are required as well:
* ~[fix handling of multiple 'active' consoles in procd](https://patchwork.ozlabs.org/project/openwrt/patch/20220125031208.17001-1-matt@traverse.com.au/) I ran into this problem after enabling framebuffer support (for VMware), causing device tree boards (QEMU virt+u-boot and Ten64) to no longer have a serial console~ Inittab restored pending rework of the console detection mechanism in procd
* 32-bit ARM kernel compilation with EFI support will fail unless [`target/linux/generic/hack-5.10/220-arm-gc_sections.patch`](https://github.com/openwrt/openwrt/blob/master/target/linux/generic/hack-5.10/220-arm-gc_sections.patch) is removed. We don't really care about kernel size here, perhaps armvirt can have a patch that reverses that one?
The errors are:
```
  LD      arch/arm/boot/compressed/vmlinux
arm-openwrt-linux-muslgnueabi-ld: warning: orphan section `.bss.efi_system_table' from `drivers/firmware/efi/libstub/ef$-stub.stub.o' being placed in section `.bss.efi_system_table'
arm-openwrt-linux-muslgnueabi-ld: warning: orphan section `.bss.flat_va_mapping' from `drivers/firmware/efi/libstub/efi$stub.stub.o' being placed in section `.bss.flat_va_mapping'    
..
arm-openwrt-linux-muslgnueabi-ld: error: zImage file size is incorrect
```

mcbridematt committed:
grub2: enable EFI for armvirt

This adds a separate package for EFI on Arm SystemReady
compatible machines. 32-bit Arm UEFI is supported as well.

It is very similar to x86-64 EFI setup, without the
need for BIOS backward compatibility and slightly
different default modules.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
build: use 128MiB as the boot/kernel partition size on armvirt target

The nominal partition type for EFI boot partitions is FAT32,
which has a minimum size of 32MiB on a 512-byte-sector block device.

To ensure that the boot partition is created as FAT32 set a size
well above this minimum.

A useful discussion about EFI partition sizes can be found here:
https://superuser.com/questions/1310927/what-is-the-absolute-minimum-size-a-uefi-system-partition-can-be

I have found 128MiB works pretty consistently across both
tools (mkfs.fat) and firmwares (EDKII)

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
build: enable vmdk/vmware images for arm64 target

This is useful for VMware's ARM64 products, e.g Fusion for M1/ARM Macs.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armvirt: add EFI support

EFI booting is used on newer machines compatible with the
Arm SystemReady specifications.

This commit restructures armvirt into a more 'generic'
target similar to x86.

See https://github.com/openwrt/openwrt/pull/4956
for a history of this port.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armvirt: update README with new image names

The introduction of EFI support has changed how armvirt
images are generated. The kernel and filesystem binaries
can still be used as before with QEMU directly.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armvirt: remove model name override

Now that armvirt has been expanded to boot on more generic
ARM machines, remove the board and model name override.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

adschm commented:
"target: armvirt:" prefix should be replaced by just "armvirt:"

mcbridematt commented:
> "target: armvirt:" prefix should be replaced by just "armvirt:"

Done.
Also changed:
* inittab restored pending rework of procd's console detection mechanism (the proposed patch potentially broke other systems)
* Add a patch that reverses 220-arm-gc-sections for this target only. I suspect some Kconfig and preprocessor tricks could fix this but I don't have enough knowledge to try that.

varoudis commented:
works like a charm on arm64/proxmox!

hauke commented on target/linux/armvirt/Makefile:
```diff
@@ -5,9 +5,9 @@
-BOARDNAME:=QEMU ARM Virtual Machine
-FEATURES:=fpu pci rtc usb
-FEATURES+=cpiogz ext4 ramdisk squashfs targz
+BOARDNAME:=Arm SystemReady and Virtual Machines
```
I would prefer to keep the old name

hauke commented on target/linux/armvirt/patches-5.10/220-reverse-arm-gc_sections-armvirt-only.patch:
```diff
@@ -0,0 +1,138 @@
+ --gc-sections"
+
+This patch conflicts with the generation of the EFI STUB for 32-bit
+ARM.
```
Probably only one or some of these strips are breaking the arm32 bit kernel. Do you know which ones?

mcbridematt replied:
It's a well above my knowledge level, but I think one or both of gc-sections and fdata-sections.
It might be that the vmlinux.ld.S for arm32 needs adjusting, as has been suggested for other platforms (arm64 https://groups.google.com/g/clang-built-linux/c/uamqmR5bHOc/m/dqOegUmKAQAJ and x86 https://www.mail-archive.com/linux-kernel@vger.kernel.org/msg1978069.html)

In the latest patchset I have reversed only the `HAVE_LD_DEAD_CODE_DATA_ELIMINATION` Kconfig instead of the entire original patch which is more manageable.

hauke commented on target/linux/armvirt/64/config-5.10:
```diff
@@ -1,150 +1,459 @@
-CONFIG_CRYPTO_SHA1_ARM64_CE=y
-CONFIG_CRYPTO_SHA256_ARM64=y
-CONFIG_CRYPTO_SHA2_ARM64_CE=y
-CONFIG_CRYPTO_SHA512_ARM64=y
```
could you split the changes in this file into multiple commits please.
Line one just activating the basic stuff to make it boot in qemu and then something adding the stuff needed for the NXP board. 
I am also surprised that you removed so many options.

mcbridematt committed:
armvirt: add ACPI support

ACPI support is required for Arm 'SystemReady' server and workstation
systems (and as an option on embedded platforms).

These config changes allow OpenWrt to boot in a QEMU virtual machine
with a UEFI/EDKII 'BIOS', but with no other hardware enabled yet.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armvirt: 64: add support for other SystemReady-compatible vendors

These changes are to support other vendors that have SystemReady/EFI
support, including:
* Marvell Armada
** (This is speculative as I don't have a machine of my own to test)
* Amazon Graviton (tested bare-metal and virtualized instances)
* VMware (Fusion for ARM Mac preview)
* NXP/Freescale (Layerscape series not already selected)
* HiSilicon
* Allwinner/sunxi
* Rockchip (untested, options taken from arm64 defconfig)

To give an idea of the hardware certified for SystemReady,
see
https://www.arm.com/architecture/system-architectures/systemready-certification-program/ir
and
https://www.arm.com/architecture/system-architectures/systemready-certification-program/es

Other vendors that _should_ work include Marvell Octeon 10
and Ampere. I understand these systems should work
"out of the box" in ACPI mode but may require other drivers
(e.g PCIe NICs and storage controllers).

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armvirt: config changes required for framebuffer console

These Kconfig options are required to get a screen console
working with the VMware Fusion ARM (Apple Silicon) preview.
They are likely to be the same for other Arm standard
"desktop" hardware that may emerge.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armvirt: base-files: add tty0 to inittab

tty0 is the default console for devices with screens/framebuffers.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt commented:
I've finally had time to update this set, I've moved it back to WIP for now.

Whats changed:

- Kernel changes for 5.15 only (must compile `CONFIG_TESTING_KERNEL=y`). If there is a solid case for 5.10 (e.g backport to 22.03 branch) I would consider doing 5.10 as well. On the other hand, is there any reason not to have armvirt on 5.15 yet?
- 32-bit EFI not supported (yet), to fix
- Split config changes into NXP DPAA2 + Other vendors (other NXP Layerscape, AWS Graviton, Marvell?, Ampere?) + VMware as suggested

Any further comments?

mcbridematt committed:
armvirt: disable LD dead code elimination on ARM32

This interferes with the generation of the EFI stub section for
ARM32. As this target is not size constrained, disable the dead code
data elimination hack.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt commented:
Updates:
* No longer WIP as 5.15 is the now armvirt default
* 32-bit EFI working again. I found I could simply deselect the (hidden) HAVE_LD_DEAD_CODE_DATA_ELIMINATION option instead of reversing the entire gc-sections patch
* Prune some dead files (5.10 configs) and unused code 
* Fix some commit titles to read 'armvirt: ' only as previously noted

hauke commented on target/linux/armvirt/config-5.15:
```diff
@@ -1,54 +1,135 @@
+CONFIG_64BIT=y
```
This will also activate `CONFIG_64BIT` on the armvirt/32 target. 
I think you should refresh the `target/linux/armvirt/config-5.15` when you select the 32 bit arm target.

mcbridematt replied:
Oops, the kernel_menuconfig for arm64 leaked back into the top level. Fixed

hauke commented on target/linux/armvirt/64/config-5.15:
```diff
@@ -134,24 +175,31 @@ CONFIG_POWER_RESET_VEXPRESS=y
 CONFIG_VEXPRESS_CONFIG=y
 CONFIG_VIDEOMODE_HELPERS=y
 CONFIG_VIRTIO_DMA_SHARED_BUFFER=y
+# CONFIG_VIRTIO_IOMMU is not set
```
This is now in generic.

hauke commented on target/linux/generic/backport-5.15/776-v5.16-net-ethernet-use-phylink_set_10g_modes.patch:
```diff
@@ -0,0 +1,66 @@
+Index: linux-5.15.45/drivers/net/ethernet/cadence/macb_main.c
+===================================================================
+--- linux-5.15.45.orig/drivers/net/ethernet/cadence/macb_main.c
++++ linux-5.15.45/drivers/net/ethernet/cadence/macb_main.c
```
Please run `make target/linux/{clean,refresh} V=99` to refresh this patch.

hauke commented on target/linux/armvirt/64/config-5.15:
```diff
@@ -99,6 +106,7 @@ CONFIG_DRM_TTM=y
 CONFIG_DRM_TTM_HELPER=y
 CONFIG_DRM_VIRTIO_GPU=y
 CONFIG_DRM_VRAM_HELPER=y
+CONFIG_ENA_ETHERNET=y
```
x86 has a kmod package for CONFIG_ENA_ETHERNET. Do you have to compile it into the kernel?
I would prefer if you move the kmod package to package/kernel/linux/modules/netdevices.mk and share it between the targets.

mcbridematt replied:
Fixed, added a change to move kmod-amazon-ena from x86 to the top level netdevices:
https://github.com/openwrt/openwrt/pull/4996/commits/1bc1899dc086534db59c9731aa4416247ec86594

hauke commented on target/linux/armvirt/64/config-5.15:
```diff
@@ -49,6 +52,7 @@ CONFIG_ARM_SMMU_V3=y
 # CONFIG_ARM_SMMU_V3_PMU is not set
 CONFIG_ATOMIC64_SELFTEST=y
 CONFIG_AUDIT_ARCH_COMPAT_GENERIC=y
+CONFIG_AQTION=y
```
Please add `kmod-atlantic` as default module to the image instead.

hauke commented on target/linux/armvirt/patches-5.15/221-armvirt-disable-gc_sections.patch:
```diff
@@ -0,0 +1,28 @@
+ 	select HAVE_UID16
+ 	select HAVE_VIRT_CPU_ACCOUNTING_GEN
+ 	select IRQ_FORCED_THREADING
+-	select HAVE_LD_DEAD_CODE_DATA_ELIMINATION
```
Do you know which sections are needed?
We could add a check to keep them for kernels with EFI.

dhewg replied:
This set has more infos:
https://lore.kernel.org/linux-arm-kernel/20181106133935.GB30658@n2100.armlinux.org.uk/
The conclusion there was that gc-sections on arm isn't worth it.

I tried it at one point, and the picture shifted a bit when comparing final compressed images, so there was still a gain. So alternatively we could replace this patch here with that unmerged set there.

On top of this, there's another hint in that thread to squeeze out a little more:
```
From: Andre Heider <a.heider@gmail.com>
Date: Mon, 6 Feb 2023 16:44:09 +0100
Subject: [PATCH] kbuild: sort sections by alignment for
 LD_DEAD_CODE_DATA_ELIMINATION

Due to the higher number of different input sections with
-f{function,data}-section, there may be lots of padding holes, since
each sections has its own alignment.

Placing larger alignments before smaller alignments can reduce the
amount of padding needed.

Signed-off-by: Andre Heider <a.heider@gmail.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 028d23841df8..27ffe3362195 100644
--- a/Makefile
+++ b/Makefile
@@ -938,7 +938,7 @@ endif
 
 ifdef CONFIG_LD_DEAD_CODE_DATA_ELIMINATION
 KBUILD_CFLAGS_KERNEL += -ffunction-sections -fdata-sections
-LDFLAGS_vmlinux += --gc-sections
+LDFLAGS_vmlinux += --gc-sections --sort-section=alignment
 endif
 
 ifdef CONFIG_SHADOW_CALL_STACK
-- 
2.39.1
```

hauke commented on target/linux/armvirt/64/target.mk:
```diff
@@ -1,8 +1,6 @@
 ARCH:=aarch64
 SUBTARGET:=64
-BOARDNAME:=QEMU ARMv8 Virtual Machine (cortex-a53)
-CPU_TYPE:=cortex-a53
```
This will switch the subtarget to generic aarch64 Is this intended?
Have you done some performance tests?

mcbridematt replied:
I'm not sure having A53 as the default CPU type is the best idea (even though our hardware has them).

There are roughly three Arm64 generations of interest: Armv8.0 (A53,A72, 2014-2017ish), Armv8.2+ (+SVE instructions, most recent Arm64 CPUs), Armv9 (+SV2 instructions, coming ~2024). 
After A55 there are no more in-order designs as well. (See https://en.wikipedia.org/wiki/Comparison_of_Armv8-A_processors)
I could try running benchmarks between the old and new AWS instances to see if choosing a specific CPU types produces an improvement.

hauke commented on target/linux/armvirt/Makefile:
```diff
@@ -6,8 +6,8 @@ include $(TOPDIR)/rules.mk
-FEATURES:=fpu pci rtc usb
-FEATURES+=cpiogz ext4 ramdisk squashfs targz
+FEATURES:=fpu pci pcie rtc usb boot-part rootfs-part
+FEATURES+=cpiogz ext4 ramdisk squashfs targz vmdk
```
I think `vmdk``support is only added in a later patch.

mcbridematt committed:
kernel: kmod-amazon-ena: move to top level netdevices

The Amazon ENA network devices are also used on the
AWS Arm (Graviton) instance types, so move it from
the x86-only module file to the top level netdevices.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

anta5010 commented on target/linux/armvirt/README:
```diff
@@ -37,4 +39,34 @@ Run with kvmtool
+    -bios QEMU_EFI_64.fd \
+    -smp 1 -m 1024 \
+    -device virtio-rng-pci \
+    -drive file=openwrt-arm-32.img,format=raw,index=0,media=disk \
```
The file name should be openwrt-arm-64.img

mcbridematt commented:
> any updates on this PR?

I'm working with another party to validate this target on other Arm SystemReady platforms (particularly ES and IR) profiles, from various other SoC vendors, more info on that soon.

anta5010 committed:
armvirt: 64: Add storage support for qemu-sbsa platform

Enable SATA support, which is used by the Server Base
System Architecture reference board[1].

Signed-off-by: Anton Antonov <Anton.Antonov@arm.com>
Signed-off-by: Mathew McBride <matt@traverse.com.au>

[1] - https://qemu.readthedocs.io/en/latest/system/arm/sbsa.html

mcbridematt committed:
armvirt: 64: add Marvell (formerly Cavium) ThunderX series network driver

Based on working configuration supplied by Anton Antonov.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armvirt: 64: disable CONFIG_SMC91X

The SMC91X family is a ISA-age Ethernet controller.
I'm not particularly sure what it's doing in armvirt/64,
as it's unlikely there is a QEMU or real hardware configuration
that exists with it.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt commented:
Hi all,

Thanks to @anta5010 (from Arm) we have been able to expand the hardware support to more SystemReady boards, including
* NXP i.MX8 Family (NXP EVKs and commercial boards from Advantech and Arduino (Portenta M8)
* Marvell (Cavium) Octeon TX2
* Ampere Altra

Some of the above (iMX8, Altra) may have incomplete network support due to firmware issues (e.g the mainline DT binding is different from the vendors reference firmware).

I have moved the network drivers into the target-specific modules file (`armvirt/modules.mk`) and there is still a bit of cleanup work there. This balances the bigger kernel size from building in the USB,MMC,PCIe and AHCI drivers for every potential platform.

We are hoping this will be adopted as the preferred OpenWrt port by Arm vendors going forward (e.g so there will not be any need to add new target/X for any Arm64 SoCs capable of booting from non-MTD storage).

Ideally it should be renamed from armvirt to something more generic (`sr_aarch64` or `sr_arm` are suggestions), but this conversation could wait until well after 23.X has been branched.

dhewg commented:
I only skimmed over this, but is it still possible to not build all the grub stuff?
Or even: do we even want grub? I may have missed the motivation for all this, but I guess this is supposed to be a "OpenWrt cloud image"? For that all that should be required is to boot a single known kernel, why is u-boot with efi not sufficient?

mcbridematt commented:
> I only skimmed over this, but is it still possible to not build all the grub stuff? Or even: do we even want grub? I may have missed the motivation for all this, but I guess this is supposed to be a "OpenWrt cloud image"? For that all that should be required is to boot a single known kernel, why is u-boot with efi not sufficient?

On most of these systems, the u-boot or EDKII is part of the system firmware so it's not OpenWrt's job to have them.
(Embedded machines with only SD cards are different, then we would bundle u-boot but have it directly hand off to the EFI bootflow)

But you are correct that GRUB isn't doing much. The main utility is being able to change the kernel command line (when needed) and determining the actual device you are booting from. (Yes, you can use the kernel EFI stub and EFI vars, but not as easy).

dhewg commented:
Okay, but the same can be said about grub. If the system firmware already has a boot loader, it's not OpenWrt's job to chain another one. OpenWrt is not a full fledged desktop distro, adding that monster just for cmdline editing seems like overkill.

I won't stop anyone who wants to use it, but can we structure this in a way that doesn't force it on everybody?
How about "Target profiles"? One to do all the grub dance, and another that doesn't?

mcbridematt commented:
> OpenWrt is not a full fledged desktop distro, adding that monster just for cmdline editing seems like overkill.

OpenWrt already has the GRUB EFI setup from x86 so it wasn't very difficult to extend that to work here.
 
> I won't stop anyone who wants to use it, but can we structure this in a way that doesn't force it on everybody?

I guess someone could either create a new device in image/Makefile that calls something other than "grub-config efi" as the first IMAGE step, or add a CONFIG_ option to generate both kinds (like BIOS and EFI on x86)

I have no objection if anyone wants to add a different EFI boot setup but I don't think it's a good idea to rework this pull request to do it.

FWIW I think it will be hard to make efistub (no grub) work on non-embedded boards as you will have to tell the kernel what the `root=` device is somehow. GRUB nicely handles that for us,

mcbridematt committed:
armvirt: 64: add Allwinner A3/A83T/A64 (sun8i family) Ethernet

Add support for the dwmac (stmmac) variant used by Allwinner
Arm64 boards.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
kernel: add mdio-bus-mux support

The MDIO bus multiplexing framework is used by some drivers
such as dwmac-sun8i.

As this is a per-driver requirement, set it to be hidden in the menu.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armvirt: package and select Rockchip DWMAC Ethernet driver

For devices that implement the "rockchip,*-gmac" compatible controller,
including:
- RK3328
- RK3399
- RK3568
- RK3588
- PX30

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
scripts: gen_image_generic: allow the partition types to be set

The use case for this is to set the kernel partition as the
EFI system partition. Versions of U-Boot with the
EFI boot manager (eficonfig and efidebug commands) will
store their boot order data on the ESP.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armvirt: set kernel partition as the EFI system partition

U-Boot with EFI boot manager functionality will store
EFI boot order data on the ESP in the ubootefi.var file.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

ShireCiel commented:
Why this merge request hangs?
It's a great change.
I am happly openwrt can run on my arm based "PVE Virtual Machine".

mcbridematt commented:
> Why this merge request hangs?
> It's a great change.
> I am happly openwrt can run on my arm based "PVE Virtual Machine".

I was about to write to OpenWrt-devel about it.
I need to do a bit more work first. At the moment the image doesn't have an EFI System Partition and that will cause issues with U-Boot's new efibootmgr.
It should be ready next week.
My current working tree is here: https://gitlab.com/traversetech/ls1088firmware/openwrt/-/commits/owrt-master-sr-2023-02-11 . Anything above "kernel: kmod-amazon-ena" will not be part of this pull request.

mcbridematt commented:
Updated pull request

Key changes:
* The kernel partition is now marked as EFI system partition (EFh)
   * requires change in gen_image_generic.sh to specify partition types - ce58b7e59fd997332ff5391143970e147ec50181
       * Have regression tested with x86 that there is no changes when partition types not specified (script/ptgen defaults to setting Linux/83h as before)
* sysupgrade will copy files, NOT dd when operating on the ESP/kernel partition
   * Avoids touching ubootefi.var - U-Boot's EFI variable persistence (see https://arm-software.github.io/ebbr/index.html#document-chapter5-variable-storage)
   * Works with 'shared storage' setups (refer [EBBR Section 4](https://arm-software.github.io/ebbr/index.html#partitioning-of-shared-storage) and https://gitlab.com/traversetech/ls1088firmware/openwrt/-/snippets/2533231 for an example HOWTO)
   * Avoids touching firmware in the ESP files, if they exist (refer [EBBR section 4.2](https://arm-software.github.io/ebbr/index.html#firmware-partition-filesystem))
* OpenWrt GRUB and kernel now live under `/efi/openwrt` with the GRUB binary also at the default/removable path of `/efi/boot/bootaa64.efi`
   * Should make it easier to do things like direct (non-GRUB) boot (via EFI vars), secure boot (`shim` in front), dual boot with non-OpenWrt, A/B setups in the future
* Similarly, sysupgrade will never change the partition table, unless `-p` is specified
   * blkid now pulled in as a dependency to resolve PARTUUID when updating grub.cfg
   * TODO (can wait): Regenerate the GRUB binary with a more precise search path at upgrade time
* Attempted to enable more platforms where EFI capable boards exist or are likely to exist
   * e.g Rockchip, Allwinner, Altera and Xilinx FPGASoCs.
* Package and select (by default) all the network device modules we know about

anta5010 committed:
armvirt: 64: Add NXP i.MX 8M Mini/Nano/Quad/Plus EVK support

Also includes Advantech RSB-3720 (iMX8 Plus) support.

Signed-off-by: Anton Antonov <Anton.Antonov@arm.com>
Signed-off-by: Mathew McBride <matt@traverse.com.au>
[Re-sort into kernel config, move network into modules]

mcbridematt committed:
armvirt: add options and driver modules for NXP Layerscape DPAA2 platform

Tested with a Traverse Technologies Ten64 (LS1088A) board.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armvirt: switch to kernel 6.1

The EFI implementation changes have only been applied
to 6.1, so switch armvirt over to it.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

john-tho commented on target/linux/armvirt/image/Makefile:
```diff
@@ -5,28 +5,113 @@
+  DEVICE_TITLE := Generic EFI Boot
+  GRUB2_VARIANT := generic
+  FILESYSTEMS := ext4 squashfs
+  DEVICE_PACKAGES += kmod-amazon-ena kmod-e1000e kmod-vmxnet3 kmod-rtc-rx8025 \
+	kmod-i2c-mux-pca954x kmod-gpio-pca953x partx-utils kmod-wdt-sp805 \
+	kmod-mvneta kmod-mvpp2 kmod-fsl-dpaa1-net kmod-fsl-dpaa2-net \
+	kmod-fsl-enetc-net kmod-dwmac-imx kmod-fsl-fec kmod-thunderx-net \
+	kmod-dwmac-rockchip kmod-dwmac-sun8i kmod-phy-aquantia kmod-phy-broadcom \
+	kmod-phy-marvell kmod-sfp
```
Thank you for pulling these out of the kernel config and into packages.
I do not know if it will apply here, but there is a qurik of the OpenWrt build system to be aware of, in that `initramfs` images will not respect `DEVICE_PACKAGES`, with [`CONFIG_BUILDBOT`](https://github.com/openwrt/openwrt/blob/4c5a9da8699a7982b8f03b28561f955d9d1313f1/scripts/target-metadata.pl#L274), specifically:
```
CONFIG_TARGET_ALL_PROFILES=y
CONFIG_TARGET_MULTI_PROFILE=y
CONFIG_TARGET_PER_DEVICE_ROOTFS=y
```
Worth doing a build with those setting (or `CONFIG_BUILDBOT`) to make sure the buildbot images will work as expected.
If you will need `initramfs` images with these drivers/packages, the workaround is to move these packages to subtarget level `DEFAULT_PACKAGES`
Example with details here: https://github.com/openwrt/openwrt/commit/1984a6bbcaac95c439dd4a6eba2f78c54e9be215


---
hauke mentioned `CONFIG_AQTION` available via `kmod-atlantic`: https://github.com/openwrt/openwrt/pull/4996#discussion_r985256853, and `CONFIG_ENA_ETHERNET` https://github.com/openwrt/openwrt/pull/4996#discussion_r985256780 is also still in the `64/config-6.1`

mcbridematt replied:
Noted, will check.
Fixed the CONFIG_ENA_ETHERNET in my local tree, will check kmod-atlantic before pushing another revision tomorrow

mcbridematt replied:
Have done a CONFIG_BUILDBOT run and all the module ko's are present in the initramfs.

mcbridematt commented:
@ynezz As discussed on OpenWrt-devel, this pull request has been updated to 6.1

Tested platforms (64-bit only except QEMU):

- Traverse Technologies Ten64 (NXP LS1088A)
- AWS Graviton t4g.nano (Neoverse-N1 VM)
- AWS Graviton bare-metal a1.metal (AWS Graviton1/A72) and c7g.metal (Graviton3/Neoverse V1)
- VMware Fusion (M1 Mac)
- QEMU 32 and 64 bit

I will evaluate the buildbot settings mentioned by @john-tho before pushing another revision shortly

ynezz commented on target/linux/armvirt/README:
```diff
@@ -37,4 +39,34 @@ Run with kvmtool
+    -netdev user,id=testwan -net nic,netdev=testwan
+
+One can find EFI/BIOS binaries from:
+- Compile mainline U-Boot for the QEMU ARM virtual machine (qemu_arm_defconfig/qemu_arm64_defconfig)
```
FYI I've [prepared](https://github.com/ynezz/openwrt/commit/bf834706e2e976c89ed732b2473f301de95f5242) `uboot-armvirt` package and it seems to be working fine.

ynezz reviewed (approved):
Just tested it, seems to be working fine in QEMU as advertised, nice work!

mcbridematt committed:
armvirt: add SFP support patches for NXP Layerscape DPAA2 platforms

This series resolves a long term issue with Phylink and SFPs on
DPAA2 hardware (LS1088,LS2088,LX2160) where a locking problem
prevented the system from rebooting. Patches solving this
issue existed prior to 6.2 but were not accepted upstream.

See the original series on patchwork:
https://patchwork.kernel.org/project/netdevbpf/cover/20221129141221.872653-1-vladimir.oltean@nxp.com/

And this thread on the Traverse forum:
https://forum.traverse.com.au/t/reboot-command-regression-from-5-10-to-5-15-kernel/133/12

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt commented:
@ynezz  This should be ready to merge now

- Fixed Amazon ENA and Aquantia hardcoded in kernel config. Created `kmod-phy-marvell-10g` as well
- Refreshed the kernel patches from the CI copies
- Fixed some missing commit descriptions, should finally pass the test formalities check (sorry for the force-push noise)
- Fix some sound-related Kconfig's not specified causing issues when CONFIG_BUILDBOT=y

I'm working on refreshing the 5.15 / 23.05 version, it should be ready early next week.

ynezz commented:
Merged in abcb30d36cfe, thanks!
